pkg/tekton/metapipeline/clientfactory.go

```diff
@@ -0,0 +1,295 @@
+// the this execution pipeline.
+func NewMetaPipelineClient() (*ClientFactory, error) {
+	tektonClient, jxClient, kubeClient, ns, err := getClientsAndNamespace()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewMetaPipelineClientWithClientsAndNamespace(jxClient, tektonClient, kubeClient, ns)
+}
+
+// NewMetaPipelineClientWithClientsAndNamespace creates a new client for the creation and application of meta pipelines using the specified parameters.
+func NewMetaPipelineClientWithClientsAndNamespace(jxClient versioned.Interface, tektonClient tektonclient.Interface, kubeClient kubernetes.Interface, ns string) (*ClientFactory, error) {
+	client := ClientFactory{
+		jxClient:       jxClient,
+		tektonClient:   tektonClient,
+		kubeClient:     kubeClient,
+		ns:             ns,
+		defaultImage:   defaultContainerImage,
+		serviceAccount: defaultServiceAccount,
+	}
+
+	return &client, nil
+}
+
+// Create creates the Tekton CRDs needed for executing the pipeline as defined by the input parameters
+func (c *ClientFactory) Create(pullRef PullRef, pipelineKind PipelineKind, context string, envs map[string]string, labels map[string]string) (kube.PromoteStepActivityKey, tekton.CRDWrapper, error) {
+	err := c.cloneVersionStreamIfNeeded()
+	if err != nil {
+		return kube.PromoteStepActivityKey{}, tekton.CRDWrapper{}, errors.Wrap(err, "unable to clone version stream")
+	}
+
+	gitInfo, err := gits.ParseGitURL(pullRef.SourceURL())
+	if err != nil {
+		return kube.PromoteStepActivityKey{}, tekton.CRDWrapper{}, errors.Wrap(err, fmt.Sprintf("unable to determine needed git info from the specified git url '%s'", pullRef.SourceURL()))
+	}
+
+	podTemplates, err := c.getPodTemplates(apps.AppPodTemplateName)
+	if err != nil {
+		return kube.PromoteStepActivityKey{}, tekton.CRDWrapper{}, errors.Wrap(err, "unable to retrieve pod templates")
+	}
+
+	branchIdentifier, err := c.determineBranchIdentifier(pipelineKind, pullRef)
+	if err != nil {
+		return kube.PromoteStepActivityKey{}, tekton.CRDWrapper{}, errors.Wrap(err, "unable to create branch identifier")
+	}
+
+	// resourceName is shared across all builds of a branch, while the pipelineName is unique for each build.
+	resourceName := tekton.PipelineResourceNameFromGitInfo(gitInfo, branchIdentifier, context, tekton.MetaPipeline, nil, "")
+	pipelineName := tekton.PipelineResourceNameFromGitInfo(gitInfo, branchIdentifier, context, tekton.MetaPipeline, c.tektonClient, c.ns)
+	buildNumber, err := tekton.GenerateNextBuildNumber(c.tektonClient, c.jxClient, c.ns, gitInfo, branchIdentifier, retryDuration, context)
+	if err != nil {
+		return kube.PromoteStepActivityKey{}, tekton.CRDWrapper{}, errors.Wrap(err, "unable to determine next build number")
+	}
+
+	logger.WithField("repo", gitInfo.URL).WithField("buildNumber", buildNumber).Info("creating meta pipeline CRDs ")
+
+	extendingApps, err := GetExtendingApps(c.jxClient, c.ns)
```

Comment:
I suppose the `GetExtendingApps` method is defined in the `metapipeline` package. Does it need to be publicly exposed from the package?

Reply:
probably not, let me check

Reply:
made private